### PR TITLE
Output executables by default

### PR DESCRIPTION
**Pull request by sbc100**
Only output object files if `-c`, `-r` or `-shared` is specified.

We have a `link_to_object` mode which was previously used by default
if the output didn't end with EXECUTABLE_ENDINGS.  After this change
`link_to_object` is only used if `-r` or `-shared` are passed
explicitly.

This allows linking executables (JS files) with arbitrary extensions.
For example, I can do `emcc foo.c -o foo.txt` and I will actually get a
JS file called `foo.txt`.  This matches the behavior of clang and gcc.

This removes some rather unhelpful errors and warnings that we were
needless generating when this `link_to_object` mode was automatically
triggered.

It also allowed me to do some refactoring to move js_target and
final_suffix declarations closer to their actual first use.

See: #11753
Replaces: #7697

### CODE REVIEW

File: tests/test_other.py [hunk 1]
```diff
@@ -409,10 +409,9 @@ def test_combining_object_files(self):
 
     # Combining object files into another object should also work, using the `-r` flag
     self.run_process([EMCC, '-r', 'twopart_main.o', 'twopart_side.o', '-o', 'combined.o'])
-    # We also support building without the `-r` flag but expect a warning
+    # Warn about legecy support for outputing object file without `-r`, `-c` or `-shared`
     err = self.run_process([EMCC, 'twopart_main.o', 'twopart_side.o', '-o', 'combined2.o'], stderr=PIPE).stderr
-    self.assertBinaryEqual('combined.o', 'combined2.o')
-    self.assertContained('warning: assuming object file output', err)
+    self.assertContained('warning: generated exectuable with object extension (.o)', err)
```
Comment (kripken) [inline tests/test_other.py:414]:
```suggestion
    self.assertContained('warning: generating am executable with an object extension (.o)', err)
```

File: emcc.py [hunk 2]
```diff
@@ -1313,14 +1316,10 @@ def check(input_file):
       if final_suffix in EXECUTABLE_ENDINGS:
         diagnostics.warning('emcc', '-shared/-r used with executable output suffix. This behaviour is deprecated.  Please remove -shared/-r to build an executable or avoid the executable suffix (%s) when building object files.' % final_suffix)
       else:
+        if options.shared:
+          diagnostics.warning('emcc', 'linking a library with `-shared` will emit a static object file.  This is form of emulation to support existing build system.  If you want to build a runtime shared library use the SIDE_MODULE setting.')
```
Comment (kripken) [inline emcc.py:1320]:
```suggestion
          diagnostics.warning('emcc', 'linking a library with `-shared` will emit a static object file.  This is a form of emulation to support existing build systems.  If you want to build a runtime shared library use the SIDE_MODULE setting.')
```
I thought I already commented about this... was it irrelevant and resolved for some reason?

File: emcc.py [hunk 3]
```diff
@@ -2041,6 +2034,9 @@ def compile_source_file(i, input_file):
         logger.debug('stopping after linking to object file')
         return 0
 
+    if final_suffix in ('.o', '.bc', '.so', '.dylib') and not shared.Settings.SIDE_MODULE:
+     diagnostics.warning('emcc', 'generated exectuable with object extension (%s).  If you meant to build an object file please use `-c, `-r`, or `-shared`' % final_suffix)
```
Comment (kripken) [inline emcc.py:2038]:
```suggestion
     diagnostics.warning('emcc', 'generating an executable with an object extension (%s).  If you meant to build an object file please use `-c, `-r`, or `-shared`' % final_suffix)
```

**Review by kripken — APPROVED** [review]

File: emcc.py [hunk 4]
```diff
@@ -1313,14 +1315,10 @@ def check(input_file):
       if final_suffix in EXECUTABLE_ENDINGS:
         diagnostics.warning('emcc', '-shared/-r used with executable output suffix. This behaviour is deprecated.  Please remove -shared/-r to build an executable or avoid the executable suffix (%s) when building object files.' % final_suffix)
       else:
+        if options.shared:
+          diagnostics.warning('emcc', 'linking a library with `-shared` will emit a static object file.  This is form of emulation to support existing building system.  If you want to build a runtime shared library use the SIDE_MODULE setting instead.')
```
[the remark this answers is not in the document]
Comment (kripken) [reply emcc.py]: Ah here, it is... is the suggestion incorrect?

[the remark this answers is not in the document]
Comment (sbc100) [reply emcc.py]: Sorry no, I simply misread.

**Pull request merged**